{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5ab695d3",
   "metadata": {},
   "source": [
    "# DATA MUNGING\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8599872",
   "metadata": {},
   "source": [
    "El objetivo de este script es, partiendo de los .csv generados por el parseador, llegar a nuevos .csv en los cuales la información esté ordenada tal y como precisan los distintos objetos generadores de figuras publicables o de resultados de análisis estadísticos. \n",
    "\n",
    "---\n",
    "No voy a hacer nada al respecto de las muestras que aparecen marcadas como sospechosas: si no han sido descartadas y han pasado el filtrado, me las creo."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98498b5b",
   "metadata": {},
   "source": [
    "Lo primero es una comprobación de que los datos a emplear sean realmente los datos definitivos vía ver si alguno de los controles internos ha amplificado y, por ende, esas muestras han de descartarse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e53bdbe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "365c7204",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were  2456 amplifications.  6  were false positives.  2450  remain.\n"
     ]
    }
   ],
   "source": [
    "filtered_data = pd.read_csv(\"../data/resistome_data/clean_data/filtered_arg\") # los arg positivos\n",
    "filtered_control = pd.read_csv(\"../data/resistome_data/clean_data/filtered_controls\") #los controles positivos\n",
    "\n",
    "# De entrada, hay mucha información que me sobra, pues para este gráfico sólo necesito las siguientes 4 variables\n",
    "filtered_data = filtered_data[[\"Sample\", \"Assay\", \"Ct\", \"chip\"]]\n",
    "filtered_control = filtered_control[[\"Sample\", \"Assay\", \"Ct\", \"chip\"]] \n",
    "\n",
    "# A continuación, voy a quitar aquellos valores con positivo en GENO vía left outer join excluyendo lo común\n",
    "before = filtered_data.shape[0]\n",
    "filtered_data = pd.merge(filtered_data, filtered_control, how = \"outer\", on = [\"Assay\", \"chip\"], indicator = True)\n",
    "common = filtered_data.loc[filtered_data[\"_merge\"] == \"both\"].shape[0]\n",
    "filtered_data = filtered_data.loc[filtered_data[\"_merge\"] == \"left_only\"]\n",
    "after = filtered_data.shape[0]\n",
    "print (\"There were \", before, \"amplifications. \", common, \" were false positives. \", after, \" remain.\")\n",
    "filtered_data.head()\n",
    "\n",
    "del before\n",
    "del common\n",
    "del after"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5705085a",
   "metadata": {},
   "source": [
    "Como se puede observar, el tamaño del df con todos los genes tras el JOIN es disinto, por lo que hemos descartado 6 amplificaciones por motivos de control interno.\n",
    "A destacar que JOIN ha cambiado el nombre de las columnas, añadiendo la terminación \"\\_x\" al df izquierdo e \"\\_y\" al derecho, por lo que toca quitarlo en cuanto tenga sentido en pos de un a mayor claridad. \n",
    "\n",
    "---\n",
    "Procedo juntar el df \"resultante\" con el comercial de la lista de genes. El método es muy similar, un join, solo que en este caso me interesa conservar lo compartido, asi que [full outer join](https://blog.codinghorror.com/a-visual-explanation-of-sql-joins/)\n",
    "\n",
    "**NOTA**: el listado de genes proviene de una hoja formato .xlxs que previamente convertí a .csv y en la cual llevé a cabo la \"limpieza\" con LibreOffice Calc, pues la información estaba dispuesta visualmente, para que fuera fácil de entender por personas, no tanto pandas. No he cambiado nada, más que reorganizar columnas y borrar cosas como el logo "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3a4c90e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Marqueur</th>\n",
       "      <th>Origine</th>\n",
       "      <th>Antibiotique</th>\n",
       "      <th>Forward</th>\n",
       "      <th>Reverse</th>\n",
       "      <th>Plaque</th>\n",
       "      <th>Puits</th>\n",
       "      <th>Informations complémentaires</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aacC2</td>\n",
       "      <td>deactivate</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>ACGGCATTCTCGATTGCTTT</td>\n",
       "      <td>CCGAGCTTCACGTAAGCATTT</td>\n",
       "      <td>Pl1</td>\n",
       "      <td>A1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aacA/aphD</td>\n",
       "      <td>deactivate</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>AGAGCCTTGGGAAGATGAAGTTT</td>\n",
       "      <td>TTGATCCATACCATAGACTATCTCATCA</td>\n",
       "      <td>Pl1</td>\n",
       "      <td>B1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(6')-II</td>\n",
       "      <td>deactivate</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>CGACCCGACTCCGAACAA</td>\n",
       "      <td>GCACGAATCCTGCCTTCTCA</td>\n",
       "      <td>Pl1</td>\n",
       "      <td>C1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aphA3</td>\n",
       "      <td>deactivate</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>AAAAGCCCGAAGAGGAACTTG</td>\n",
       "      <td>CATCTTTCACAAAGATGTTGCTGTCT</td>\n",
       "      <td>Pl1</td>\n",
       "      <td>D1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sat4</td>\n",
       "      <td>deactivate</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>GAATGGGCAAAGCATAAAAACTTG</td>\n",
       "      <td>CCGATTTTGAAACCACAATTATGATA</td>\n",
       "      <td>Pl1</td>\n",
       "      <td>E1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Marqueur     Origine    Antibiotique                   Forward  \\\n",
       "0       aacC2  deactivate  Aminoglycoside      ACGGCATTCTCGATTGCTTT   \n",
       "1   aacA/aphD  deactivate  Aminoglycoside   AGAGCCTTGGGAAGATGAAGTTT   \n",
       "2  aac(6')-II  deactivate  Aminoglycoside        CGACCCGACTCCGAACAA   \n",
       "3       aphA3  deactivate  Aminoglycoside     AAAAGCCCGAAGAGGAACTTG   \n",
       "4        sat4  deactivate  Aminoglycoside  GAATGGGCAAAGCATAAAAACTTG   \n",
       "\n",
       "                        Reverse Plaque Puits Informations complémentaires  \n",
       "0         CCGAGCTTCACGTAAGCATTT    Pl1    A1                          NaN  \n",
       "1  TTGATCCATACCATAGACTATCTCATCA    Pl1    B1                          NaN  \n",
       "2          GCACGAATCCTGCCTTCTCA    Pl1    C1                          NaN  \n",
       "3    CATCTTTCACAAAGATGTTGCTGTCT    Pl1    D1                          NaN  \n",
       "4    CCGATTTTGAAACCACAATTATGATA    Pl1    E1                          NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arg_names = pd.read_csv(\"../data/resistome_data/metadada/a_names.csv\", sep = \",\")\n",
    "arg_names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9157316d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay</th>\n",
       "      <th>antib</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>dfrAB4</td>\n",
       "      <td>trimethoprim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>dfrC</td>\n",
       "      <td>trimethoprim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>dfrG</td>\n",
       "      <td>trimethoprim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>dfrK</td>\n",
       "      <td>trimethoprim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>dfrBmulti</td>\n",
       "      <td>trimethoprim</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Assay         antib\n",
       "377     dfrAB4  trimethoprim\n",
       "378       dfrC  trimethoprim\n",
       "379       dfrG  trimethoprim\n",
       "380       dfrK  trimethoprim\n",
       "381  dfrBmulti  trimethoprim"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Nuevamente, del df para hacer el gráfico me interesa conservar la información justa y necesaria\n",
    "arg_names = arg_names[[\"Marqueur\", \"Antibiotique\"]]\n",
    "arg_names.rename(columns = {\"Marqueur\": \"Assay\", \"Antibiotique\": \"antib\"}, inplace = True) # Una cosa es programar en spanglish y otra en francés\n",
    "arg_names.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "036b27e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Assay</th>\n",
       "      <th>Ct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>aph4ib</td>\n",
       "      <td>26.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>aph4ib</td>\n",
       "      <td>26.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>aph4ib</td>\n",
       "      <td>25.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>spcN</td>\n",
       "      <td>26.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>spcN</td>\n",
       "      <td>26.21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample   Assay     Ct\n",
       "0       7  aph4ib  26.05\n",
       "1       8  aph4ib  26.36\n",
       "2       9  aph4ib  25.33\n",
       "3       7    spcN  26.89\n",
       "4       9    spcN  26.21"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Antes de hacer nada, reorganizar filtered_data quitando lo sobrante\n",
    "filtered_data.drop(columns = [\"Sample_y\", \"Ct_y\", \"chip\", \"_merge\"], axis = 1, inplace = True) #chip sobra, basta con sample. _merge ya no nos es útil\n",
    "filtered_data.rename(columns = {\"Sample_x\": \"sample\", \"Ct_x\": \"Ct\"}, inplace = True)\n",
    "filtered_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7af70d58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay</th>\n",
       "      <th>antib</th>\n",
       "      <th>sample</th>\n",
       "      <th>Ct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2481</th>\n",
       "      <td>blaB-11,13,14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>24.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2482</th>\n",
       "      <td>blaB-11,13,14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21.0</td>\n",
       "      <td>24.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2483</th>\n",
       "      <td>blaB-11,13,14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "      <td>22.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2484</th>\n",
       "      <td>pBS228-IncP-1?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>21.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2485</th>\n",
       "      <td>pBS228-IncP-1?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21.0</td>\n",
       "      <td>20.57</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Assay antib  sample     Ct\n",
       "2481   blaB-11,13,14   NaN    20.0  24.33\n",
       "2482   blaB-11,13,14   NaN    21.0  24.16\n",
       "2483   blaB-11,13,14   NaN    23.0  22.58\n",
       "2484  pBS228-IncP-1?   NaN    20.0  21.58\n",
       "2485  pBS228-IncP-1?   NaN    21.0  20.57"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Y ya por fin, llevar a cabo el join\n",
    "full_data = pd.merge(arg_names, filtered_data, how = \"outer\", on = [\"Assay\"])\n",
    "full_data.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a21038b4",
   "metadata": {},
   "source": [
    "En ese .tail() se puede observar un hecho bastante preocupante: ¿qué hacen ahí dos amplificaciones de GENO? No las había pillado hasta ahora porque se encuentran en \"Assay\", en la columna que contiene los nombres de los ARG testeados, y la verdad el que estén ahí suscita preguntas sobre el diseño del chip que se me pueden estar escapando (recordemos que el cómo está diseñado el chip y cómo analizar sus resultados lo he sacado por ingeniería inversa, que los del servicio externo cobraban como un extra aparte el análisis y por ende esa información tan esencial no la comparten)\n",
    "\n",
    "<mark>NOTA: están dentro de muestras marcadas como 000NTC. No entiendo nada, las ignoro</mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1a340d39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2452\n",
      "['Aminoglycoside' 'Amphenicol' 'Vancomycin' 'Beta Lactam' 'beta Lactam'\n",
      " 'Beta lactamase' 'MDR' 'MDR-chromo' 'other' 'Fluoroquinolone'\n",
      " 'fluoroquinolone' 'MLSB' 'Phenicol' 'Tetracycline' 'Tetracycline '\n",
      " 'Insertional' 'Insertional ' 'Integrase' 'MDR-mobile' 'MGE' 'plasmid'\n",
      " 'Plasmid-inc' 'Plasmid-inc ' 'Plasmid-rep' 'Sulfonamide' 'Transposase'\n",
      " 'trimethoprim']\n"
     ]
    }
   ],
   "source": [
    "full_data = full_data[~full_data[\"Assay\"].isin([\"GENO\"])] #también vuelven a escena los 16S por meter los Assays\n",
    "full_data = full_data[~pd.isnull(full_data[\"antib\"])]\n",
    "full_data = full_data[(full_data[\"antib\"] != \"16S\")]\n",
    "print(full_data.shape[0])\n",
    "print(full_data[\"antib\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beebfd3e",
   "metadata": {},
   "source": [
    "Como se puede ver en la lista de \"antibióticos\" (que en realidad también incluye MGE), hay muchos nombres mal puestos, así como valores de 16S que habría que quitar antes de seguir. Por tanto, hay que limpiar antes de seguir adelante"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bc6d1ca7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Aminoglycoside' 'Amphenicol' 'Vancomycin' 'Beta Lactam' 'MDR' 'Other'\n",
      " 'Fluoroquinolone' 'MLSB' 'Phenicol' 'Tetracycline' 'Insertional'\n",
      " 'Integrase' 'MGE' 'Plasmid' 'Plasmid-inc' 'Plasmid-rep' 'Sulfonamide'\n",
      " 'Transposase' 'Trimethoprim']\n"
     ]
    }
   ],
   "source": [
    "full_data.loc[(full_data[\"antib\"] == \"MDR-chromo\") | (full_data[\"antib\"] == \"MDR-mobile\"), \"antib\"] = \"MDR\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"Beta lactamase\") | (full_data[\"antib\"] == \"beta Lactam\"), \"antib\"] = \"Beta Lactam\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"trimethoprim\"), \"antib\"] = \"Trimethoprim\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"Tetracycline \"), \"antib\"] = \"Tetracycline\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"fluoroquinolone\"), \"antib\"] = \"Fluoroquinolone\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"other\"), \"antib\"] = \"Other\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"Insertional \"), \"antib\"] = \"Insertional\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"Plasmid-inc \"), \"antib\"] = \"Plasmid-inc\"\n",
    "full_data.loc[(full_data[\"antib\"] == \"plasmid\"), \"antib\"] = \"Plasmid\"\n",
    "\n",
    "print(full_data[\"antib\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4ac3b4a",
   "metadata": {},
   "source": [
    "Un último apaño que puedo hacer en este punto es desglosar la información contenida en \"sample\": yo sé que el sample 1 se corresponde con la reṕlica 1 de **suelo** del **uruguay**, por ejemplo. Asi que podría crear columnas que indicaran el tipo general de muestra (plástico/control), su ubicación (uruguay/ionosférico) e incluso su tipo específico (EPS/PUR/Agua/Suelo), para así poder agrupar mucho más fácilmente de cara a futuros análisis/representaciones. Para ello, es útil tener esto a mano:\n",
    "\n",
    "chip | sample | Desc.\n",
    "---:|:---:|---\n",
    "**1** | 1 | URU Suelo 1\n",
    "// | 2 | URU Suelo 2\n",
    "// | 3 | URU Suelo 3\n",
    "**2** | 4 | URU Agua 1\n",
    "// | 5 | URU Agua 2\n",
    "// | 6 | URU Agua 3\n",
    "**3** | 7 | URU EPS 1\n",
    "// | 8 | URU EPS 2\n",
    "// | 9 | URU EPS 3\n",
    "**4** | 10 | URU PUR 1\n",
    "// | 11 | URU PUR 2\n",
    "// | 12 | URU PUR 3\n",
    "**5** | 13 | IÓN Suelos 1\n",
    "// | 14 | IÓN Agua 1\n",
    "// | 15 | IÓN EPS 1\n",
    "**6** | 16 | IÓN EPS 2\n",
    "// | 17 | IÓN PUR 1\n",
    "// | 18 | IÓN PUR 2\n",
    "**7** | 19 | ARDLEY Suelos 1\n",
    "// | 20 | ARDLEY Suelos 2\n",
    "// | 21 | ARDLEY Suelos 3\n",
    "**8** | 22 | ARDLEY EPS 1\n",
    "// | 23 | ARDLEY EPS 2\n",
    "// | 24 | ARDLEY EPS 3\n",
    "**9** | 25 | ARDLEY PUR 1\n",
    "// | 26 | ARDLEY PUR 2\n",
    "// | 27 | ARDLEY PUR 3\n",
    "\n",
    "Hay otra clasificación más que podemos hacer aquí: entre ARG y MGE, pues ahora todo está en la columna \"antib\" y nos interesará separar en un futuro. Aquí la lista de ARG / MGE:\n",
    "- **ARG**: Aminoglycoside, Amphenicol, Vancomycin, Beta Lactam, MDR, Fluoroquinolone, MLSB, Phenicol, Tetracycline, Sulfonamide, Trimethoprim\n",
    "- **MGE**: MGE, Insertional, Plasmid, Plasmid-inc, Plasmid-rep, Transposase, Other, Integrase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "63bf6d2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Assay         antib  sample     Ct   place type_f   type_g or_seq\n",
      "2449  dfrBmulti  Trimethoprim    26.0  15.72  ardley    PUR  plastic    arg\n",
      "2450  dfrBmulti  Trimethoprim    27.0  15.77  ardley    PUR  plastic    arg\n",
      "2451  dfrBmulti  Trimethoprim    22.0  16.84  ardley    EPS  plastic    arg\n",
      "2452  dfrBmulti  Trimethoprim    23.0  17.13  ardley    EPS  plastic    arg\n",
      "2453  dfrBmulti  Trimethoprim    24.0  17.15  ardley    EPS  plastic    arg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_9382/2864138303.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'uru' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  full_data.loc[(full_data[\"sample\"] < 13), \"place\"] = \"uru\"\n",
      "/tmp/ipykernel_9382/2864138303.py:7: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'soil' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  full_data.loc[(full_data[\"sample\"].isin([1, 2, 3, 13, 19, 20, 21])), \"type_f\"] = \"soil\"\n",
      "/tmp/ipykernel_9382/2864138303.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'plastic' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  full_data.loc[(full_data[\"type_f\"] == \"EPS\") | (full_data[\"type_f\"] == \"PUR\"), \"type_g\"] = \"plastic\"\n",
      "/tmp/ipykernel_9382/2864138303.py:20: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'arg' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  full_data.loc[(full_data[\"antib\"].isin(arg_lst)), \"or_seq\"] = \"arg\"\n"
     ]
    }
   ],
   "source": [
    "#primero, definir puntos de muestreo\n",
    "full_data.loc[(full_data[\"sample\"] < 13), \"place\"] = \"uru\"\n",
    "full_data.loc[(full_data[\"sample\"] > 12) & (full_data[\"sample\"] < 19), \"place\"] = \"ion\"\n",
    "full_data.loc[(full_data[\"sample\"] > 18), \"place\"] = \"ardley\"\n",
    "\n",
    "#segundo, definir tipo fino\n",
    "full_data.loc[(full_data[\"sample\"].isin([1, 2, 3, 13, 19, 20, 21])), \"type_f\"] = \"soil\"\n",
    "full_data.loc[(full_data[\"sample\"].isin([4, 5, 6, 14])), \"type_f\"] = \"water\"\n",
    "full_data.loc[(full_data[\"sample\"].isin([7, 8, 9, 15, 16, 22, 23, 24])), \"type_f\"] = \"EPS\"\n",
    "full_data.loc[(full_data[\"sample\"].isin([10, 11, 12, 17, 18, 25, 26, 27])), \"type_f\"] = \"PUR\"\n",
    "\n",
    "#tercero, tipo general\n",
    "full_data.loc[(full_data[\"type_f\"] == \"EPS\") | (full_data[\"type_f\"] == \"PUR\"), \"type_g\"] = \"plastic\"\n",
    "full_data.loc[(full_data[\"type_f\"] == \"water\") | (full_data[\"type_f\"] == \"soil\"), \"type_g\"] = \"control\"\n",
    "\n",
    "#Y por último, ARG vs MGE\n",
    "arg_lst = [\"Aminoglycoside\", \"Amphenicol\", \"Vancomycin\", \"Beta Lactam\", \"MDR\", \n",
    "           \"Fluoroquinolone\", \"MLSB\", \"Phenicol\", \"Tetracycline\", \"Sulfonamide\", \n",
    "           \"Trimethoprim\"]\n",
    "full_data.loc[(full_data[\"antib\"].isin(arg_lst)), \"or_seq\"] = \"arg\"\n",
    "full_data.loc[(~full_data[\"antib\"].isin(arg_lst)), \"or_seq\"] = \"mge\"\n",
    "\n",
    "#y chequeo rápido\n",
    "check_data = full_data.dropna()\n",
    "print(check_data.tail())\n",
    "del check_data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "334d526e",
   "metadata": {},
   "source": [
    "Aquí hay una primera divergencia en la biliografía.\n",
    "- Por un lado, hay trabajos como [Jiang et al, 2023](https://www.sciencedirect.com/science/article/pii/S0048969723007556) y [Do et al, 2023](https://www.biorxiv.org/content/10.1101/2022.02.25.481976v1.abstract) (ojo que este es una preprint) que proceden a calcular el nº de copia relativo (desarrollo abajo) y con él, lo dividen entre el nº de copia del 16S para normalizarlo y analizan los resultados en base a esa nueva variable. \n",
    "- Por otro lado, hay trabajos PON EJEMPLOS en los que para comparar la expresión relativa entre condiciones de estudio se calcula el fold-change vía delta delta Ct.\n",
    "- \\* hay incluso papers en los que calculan las dos cosas y luego no especifican cuál emplean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b46675e",
   "metadata": {},
   "source": [
    "---\n",
    "De primeras voy con el primer método, el de [Jiang et al, 2023](https://www.sciencedirect.com/science/article/pii/S0048969723007556).\n",
    "\n",
    "En él, se hace uso de la [fórmula de Looft et al, 2012](https://www.pnas.org/doi/abs/10.1073/pnas.1120238109) para obtener _RN_, el nº de copias relativo:\n",
    "\n",
    "$$\n",
    "RN = 10^{\\frac{27-Ct}{10/3}}\n",
    "$$\n",
    "Tomamos 27 como límite de detección (Looft y Jiang emplearon distintos límites y por ende su fórmula es distinta)\n",
    "\n",
    "\n",
    "RN se obtiene tanto para cada ARGvMGE y para el gen de housekeeping (en nuestro caso, el 16S), para finalmente normalizar el uno con el otro. En nuestro caso:\n",
    "\n",
    "$$\n",
    "RA = \\frac{RN_{ARG}}{RN_{16S}}\n",
    "$$\n",
    "\n",
    "Siendo _RA_ la abundancia relativa (_Relative Abundance_), la variable que emplearemos en los análisis estadísticos de ahora en adelante. Este procedimiento es de Looft entero, ojo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ffae672b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    23.78\n",
       "1    22.51\n",
       "2    25.91\n",
       "3    24.56\n",
       "4    26.05\n",
       "Name: Ct, dtype: float64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_data[\"rel_n\"] = (10 ** ((27 - full_data[\"Ct\"])/(10/3))) #esto añade el valor\n",
    "#full_data.head()\n",
    "check_df = full_data[\"Ct\"].dropna() # chequeo de esta manera pq si no sólo vería valores de Ct = 0\n",
    "full_data[\"rel_n\"].fillna(0, inplace = True) #para poder seguir operando\n",
    "check_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a3a646e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "del check_df #ya ha cumplido su función"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcf4db40",
   "metadata": {},
   "source": [
    "Una vez se tiene el nº de copia relativo, lo siguiente es obtener el \"absoluto\" dividiendo el nº de copia relativo de cada gen entre el nº de copia relativo del 16S. O lo que es lo mismo, toca abrir los datos del 16S.\n",
    "\n",
    "Como viene siendo el caso, no bastará con abrirlos y punto, sino que tendré que quedarme sólo con la información de interés (quitando columnas de más y los 000NTC y GENO), así como ver qué hacer en aquellos casos en los que uno de los valores no sirva. Lo suyo sería usar siempre el mismo gen de 16S de los dos, o al menos el mismo para cada muestra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5550a67e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['7' '8' '9' '25' '26' '27' '13' '14' '15' '10' '11' '12' '4' '5' '6' '1'\n",
      " '2' '3' '19' '20' '22' '23' '24' '16' '17' '18']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Assay</th>\n",
       "      <th>Ct</th>\n",
       "      <th>chip</th>\n",
       "      <th>Efficiency</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>1</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>12.22</td>\n",
       "      <td>1</td>\n",
       "      <td>2.24</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>1</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.58</td>\n",
       "      <td>1</td>\n",
       "      <td>1.68</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>10</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.18</td>\n",
       "      <td>4</td>\n",
       "      <td>1.87</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>11</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.22</td>\n",
       "      <td>4</td>\n",
       "      <td>1.88</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>12</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.73</td>\n",
       "      <td>4</td>\n",
       "      <td>1.90</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>12</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>12.39</td>\n",
       "      <td>4</td>\n",
       "      <td>2.08</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>13</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>14.40</td>\n",
       "      <td>5</td>\n",
       "      <td>1.92</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>14</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>15.42</td>\n",
       "      <td>5</td>\n",
       "      <td>1.89</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>15</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.29</td>\n",
       "      <td>5</td>\n",
       "      <td>1.88</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>16</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.64</td>\n",
       "      <td>6</td>\n",
       "      <td>1.84</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>16</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>12.15</td>\n",
       "      <td>6</td>\n",
       "      <td>2.13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>17</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>11.83</td>\n",
       "      <td>6</td>\n",
       "      <td>2.16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>17</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.24</td>\n",
       "      <td>6</td>\n",
       "      <td>1.82</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>18</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.21</td>\n",
       "      <td>6</td>\n",
       "      <td>1.79</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>19</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>11.47</td>\n",
       "      <td>7</td>\n",
       "      <td>2.01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>2</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>12.78</td>\n",
       "      <td>1</td>\n",
       "      <td>1.78</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>2</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>11.50</td>\n",
       "      <td>1</td>\n",
       "      <td>2.02</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>20</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>8.84</td>\n",
       "      <td>7</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>20</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>12.05</td>\n",
       "      <td>7</td>\n",
       "      <td>2.06</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>22</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>12.29</td>\n",
       "      <td>8</td>\n",
       "      <td>2.01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>22</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>8.33</td>\n",
       "      <td>8</td>\n",
       "      <td>1.87</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>23</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>8.30</td>\n",
       "      <td>8</td>\n",
       "      <td>1.81</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>23</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>12.43</td>\n",
       "      <td>8</td>\n",
       "      <td>1.96</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>24</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>12.56</td>\n",
       "      <td>8</td>\n",
       "      <td>1.99</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>24</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>8.77</td>\n",
       "      <td>8</td>\n",
       "      <td>1.90</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>25</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>10.99</td>\n",
       "      <td>9</td>\n",
       "      <td>1.96</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>25</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>7.61</td>\n",
       "      <td>9</td>\n",
       "      <td>2.29</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>26</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>11.08</td>\n",
       "      <td>9</td>\n",
       "      <td>1.99</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>27</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>7.49</td>\n",
       "      <td>9</td>\n",
       "      <td>2.21</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>27</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>11.42</td>\n",
       "      <td>9</td>\n",
       "      <td>2.14</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>3</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.25</td>\n",
       "      <td>1</td>\n",
       "      <td>1.73</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>3</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>12.23</td>\n",
       "      <td>1</td>\n",
       "      <td>2.27</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>4</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>15.64</td>\n",
       "      <td>2</td>\n",
       "      <td>1.90</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>5</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>14.94</td>\n",
       "      <td>2</td>\n",
       "      <td>1.90</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>6</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>16.28</td>\n",
       "      <td>2</td>\n",
       "      <td>1.90</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>12.93</td>\n",
       "      <td>3</td>\n",
       "      <td>2.02</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.83</td>\n",
       "      <td>3</td>\n",
       "      <td>1.90</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>12.66</td>\n",
       "      <td>3</td>\n",
       "      <td>2.03</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>13.53</td>\n",
       "      <td>3</td>\n",
       "      <td>1.89</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>16S old 1</td>\n",
       "      <td>14.04</td>\n",
       "      <td>3</td>\n",
       "      <td>1.88</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9</td>\n",
       "      <td>16S new 2</td>\n",
       "      <td>13.28</td>\n",
       "      <td>3</td>\n",
       "      <td>1.95</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample      Assay     Ct  chip  Efficiency  N\n",
       "55      1  16S new 2  12.22     1        2.24  3\n",
       "50      1  16S old 1  13.58     1        1.68  2\n",
       "30     10  16S old 1  13.18     4        1.87  3\n",
       "31     11  16S old 1  13.22     4        1.88  3\n",
       "32     12  16S old 1  13.73     4        1.90  3\n",
       "37     12  16S new 2  12.39     4        2.08  3\n",
       "20     13  16S old 1  14.40     5        1.92  3\n",
       "21     14  16S old 1  15.42     5        1.89  3\n",
       "22     15  16S old 1  13.29     5        1.88  3\n",
       "80     16  16S old 1  13.64     6        1.84  3\n",
       "85     16  16S new 2  12.15     6        2.13  2\n",
       "86     17  16S new 2  11.83     6        2.16  3\n",
       "81     17  16S old 1  13.24     6        1.82  3\n",
       "82     18  16S old 1  13.21     6        1.79  3\n",
       "60     19  16S old 1  11.47     7        2.01  3\n",
       "51      2  16S old 1  12.78     1        1.78  2\n",
       "56      2  16S new 2  11.50     1        2.02  2\n",
       "66     20  16S new 2   8.84     7        2.00  2\n",
       "61     20  16S old 1  12.05     7        2.06  3\n",
       "70     22  16S old 1  12.29     8        2.01  3\n",
       "75     22  16S new 2   8.33     8        1.87  3\n",
       "76     23  16S new 2   8.30     8        1.81  3\n",
       "71     23  16S old 1  12.43     8        1.96  3\n",
       "72     24  16S old 1  12.56     8        1.99  3\n",
       "77     24  16S new 2   8.77     8        1.90  3\n",
       "10     25  16S old 1  10.99     9        1.96  2\n",
       "15     25  16S new 2   7.61     9        2.29  2\n",
       "11     26  16S old 1  11.08     9        1.99  1\n",
       "17     27  16S new 2   7.49     9        2.21  2\n",
       "12     27  16S old 1  11.42     9        2.14  3\n",
       "52      3  16S old 1  13.25     1        1.73  3\n",
       "57      3  16S new 2  12.23     1        2.27  3\n",
       "40      4  16S old 1  15.64     2        1.90  3\n",
       "41      5  16S old 1  14.94     2        1.90  3\n",
       "42      6  16S old 1  16.28     2        1.90  3\n",
       "5       7  16S new 2  12.93     3        2.02  3\n",
       "0       7  16S old 1  13.83     3        1.90  3\n",
       "6       8  16S new 2  12.66     3        2.03  3\n",
       "1       8  16S old 1  13.53     3        1.89  3\n",
       "2       9  16S old 1  14.04     3        1.88  3\n",
       "7       9  16S new 2  13.28     3        1.95  3"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arnr_data = pd.read_csv(\"../data/resistome_data/clean_data/16S_refs\")\n",
    "arnr_data = arnr_data[[\"Sample\", \"Assay\", \"Ct\", \"chip\", \"Efficiency\", \"N\"]] #dejo la eficiencia y N para poder chequear qué valores son los apropiados\n",
    "arnr_data = arnr_data.loc[~arnr_data[\"Sample\"].isin([\"000NTC\", \"GENO\"])] #quito esto\n",
    "arnr_data.dropna(inplace = True)\n",
    "print(arnr_data[\"Sample\"].unique()) #para chequear que haya amplificación en todas las muestras\n",
    "arnr_data.sort_values(by = \"Sample\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "134d2afe",
   "metadata": {},
   "source": [
    "En base a lo que se observa en el df, aunque hay amplificación en todas las muestras, fallan dos detalles:\n",
    "- No hay suficiente 16S new para usarlo con todas las muestras, pero sí 16S old 1\n",
    "- Si filtro por eficiencia como hago de normal, pierdo la muestra 1 y 3, pero por muy, muy poco tanto en new como en old\n",
    "Por tanto, por ahora mantengo el 16S old 1, aunque \"falle\" en dos muestras, y uso esos valores "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "21f07b58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{7: 8933.054837332944,\n",
       " 8: 10990.05839432522,\n",
       " 9: 7726.805850957021,\n",
       " 25: 63533.09318517423,\n",
       " 26: 59703.52865838366,\n",
       " 27: 47206.304126359,\n",
       " 13: 6025.595860743575,\n",
       " 14: 2978.516429429188,\n",
       " 15: 12971.792709839572,\n",
       " 10: 13995.873225726178,\n",
       " 11: 13614.446824659482,\n",
       " 12: 9571.94071294844,\n",
       " 4: 2558.585886905643,\n",
       " 5: 4149.5404263436285,\n",
       " 6: 1644.371723214929,\n",
       " 1: 10616.955571987242,\n",
       " 2: 18450.154191794736,\n",
       " 3: 13335.21432163324,\n",
       " 19: 45603.691595129596,\n",
       " 20: 30549.21113215509,\n",
       " 22: 25882.12915153093,\n",
       " 23: 23496.32820848305,\n",
       " 24: 21478.304741305332,\n",
       " 16: 10185.91388054117,\n",
       " 17: 13427.649611378642,\n",
       " 18: 13708.817661648523}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arnr_data = arnr_data.loc[~arnr_data[\"Assay\"].isin([\"16S new 2\"])] #dejar sólo 16S old 1\n",
    "#print(arnr_data[\"Sample\"].unique()) #para chequear que siga siendo útil\n",
    "\n",
    "arnr_data[\"rel_n\"] = (10 ** ((27 - arnr_data[\"Ct\"])/(10/3))) #para obtener el nº de copia relativo del 16S\n",
    "arnr_data = arnr_data[[\"Sample\", \"rel_n\"]] # sólo lo provechoso\n",
    "arnr_data[\"Sample\"] = pd.to_numeric(arnr_data[\"Sample\"], errors = \"coerce\")\n",
    "arnr_data.set_index(\"Sample\", inplace = True) # necesario por cómo funciona la función .to_dict()\n",
    "arnr_vals = arnr_data.to_dict()[\"rel_n\"] #aquí el diccionario\n",
    "del arnr_data #ya ha cumplido su función\n",
    "arnr_vals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51a0a088",
   "metadata": {},
   "source": [
    "Y ya obtener el valor de abundancia relativa (RA) _per se_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d60cf006",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Assay         antib  sample     Ct   place type_f   type_g or_seq  \\\n",
      "2449  dfrBmulti  Trimethoprim    26.0  15.72  ardley    PUR  plastic    arg   \n",
      "2450  dfrBmulti  Trimethoprim    27.0  15.77  ardley    PUR  plastic    arg   \n",
      "2451  dfrBmulti  Trimethoprim    22.0  16.84  ardley    EPS  plastic    arg   \n",
      "2452  dfrBmulti  Trimethoprim    23.0  17.13  ardley    EPS  plastic    arg   \n",
      "2453  dfrBmulti  Trimethoprim    24.0  17.15  ardley    EPS  plastic    arg   \n",
      "\n",
      "            rel_n        rel_16    rel_ab  \n",
      "2449  2421.029047  59703.528658  0.040551  \n",
      "2450  2338.837239  47206.304126  0.049545  \n",
      "2451  1116.863248  25882.129152  0.043152  \n",
      "2452   914.113241  23496.328208  0.038905  \n",
      "2453   901.571138  21478.304741  0.041976  \n"
     ]
    }
   ],
   "source": [
    "#si la key (nº muestra) es igual al de un valor dado, asigna el 16S correspondiente\n",
    "for sample_num,rel_16 in arnr_vals.items():\n",
    "    full_data.loc[(full_data[\"sample\"] == sample_num), \"rel_16\"] = rel_16 \n",
    "full_data[\"rel_ab\"] = full_data[\"rel_n\"]/full_data[\"rel_16\"] \n",
    "\n",
    "check_df = full_data.dropna() #nuevamente, para chequear y poder ver algo que no sea NaN o 0\n",
    "print(check_df.tail())\n",
    "del check_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc315f2c",
   "metadata": {},
   "source": [
    "Un problema que ha surgido al pasar de Ct a la abundancia relativa es que los valores de abundancia son ínfimos: apenas hay. No sólo eso, sino que también se da el caso que las diferencias entre ellos son muy grandes (al fin y al cabo, las diferencias de unidades de Cts se transforman en distancias exponenciales al pasar a nº relativo de copia). Para solucionarlo, empleo una transformación logarítmica"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f49ccbf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Assay           antib  sample     Ct   place type_f   type_g or_seq  \\\n",
      "0      aacC2  Aminoglycoside    25.0  23.78  ardley    PUR  plastic    arg   \n",
      "1      aacC2  Aminoglycoside    27.0  22.51  ardley    PUR  plastic    arg   \n",
      "2      aacC2  Aminoglycoside    20.0  25.91  ardley   soil  control    arg   \n",
      "3      aacC2  Aminoglycoside    22.0  24.56  ardley    EPS  plastic    arg   \n",
      "4  aacA/aphD  Aminoglycoside    25.0  26.05  ardley    PUR  plastic    arg   \n",
      "\n",
      "       rel_n        rel_16    rel_ab  log_n  \n",
      "0   9.246982  63533.093185  0.000146 -3.837  \n",
      "1  22.233099  47206.304126  0.000471 -3.327  \n",
      "2   2.123244  30549.211132  0.000070 -4.158  \n",
      "3   5.395106  25882.129152  0.000208 -3.681  \n",
      "4   1.927525  63533.093185  0.000030 -4.518  \n"
     ]
    }
   ],
   "source": [
    "## Voy a calcular el log10 de manera chapucera, pues al calcularlo a la vez en todas las casillas de la columna\n",
    "## y tener la mayoría de casillas rel_ab = NaN, lo que ocurrirá en la mayoría de casos es que el programa detectará\n",
    "## automáticamente que le estoy pidiendo una imposibilidad matemática y simplemente se saltará ese cálculo.\n",
    "## A efectos prácticos, el resultado final será el mismo, pues debiera valer 0 y le haré valer 0\n",
    "full_data[\"log_n\"] = np.log10(full_data[\"rel_ab\"])\n",
    "full_data.loc[(full_data[\"rel_ab\"] == 0.0), \"log_n\"] = 0\n",
    "\n",
    "\n",
    "#como de costumbre ya\n",
    "check_df = full_data.dropna() #nuevamente, para chequear y poder ver algo que no sea NaN o 0\n",
    "print(check_df.head())\n",
    "del check_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b577c436",
   "metadata": {},
   "source": [
    "Llegados a este punto, ya tengo un df útil para hacer gráficos. En particular, para hacer diagramas de venn de \n",
    "presencia/ausencia de ARG/MGE,  diagramas de barras (o similares) de abundancia relativa y cálculos estadísticos.\n",
    "Por tanto, guardo el df como .csv con el nombre **\"ab_data_simple.csv\"** (ab de abundancia relativa, data de data y simple porque es la versión mínima con toda la información posible)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8ed9c2e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#full_data.fillna(0, inplace = True)\n",
    "full_data.to_csv(\"../data/resistome_data/clean_data/ab_data_simple.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c5c41cb",
   "metadata": {},
   "source": [
    "---\n",
    "Hasta este momento he estado trabajando con datos \"mínimos\", véase, con un dataset en el que estuviera representada la información mínima necesaria para entender el global. Sin embargo, para según qué representaciones (de entrada, el heatmap) necesitaré tener en cuenta TODA la información generada: un df en el que se recoja la amplificación/falta de de todos los ARG y MGE **de todas las muestras**. \n",
    "\n",
    "Dado que tengo 18 muestras (este nº cambiará cuando reciba todos los resultados) y en torno a 429 arg/mge, debiera tener un df final de 6876 filas. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "277ed210",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9932\n",
      "                  Assay          antib_x          antib_y  sample     Ct  \\\n",
      "194            qepA_1_2  Fluoroquinolone  Fluoroquinolone    14.0  27.00   \n",
      "283               tetPA    Tetracycline      Tetracycline    14.0  26.41   \n",
      "299  intI1F165_clinical        Integrase        Integrase    14.0  25.19   \n",
      "308                czcA       MDR-mobile              MDR    14.0  21.77   \n",
      "340               trb-C          plasmid          Plasmid    14.0  26.80   \n",
      "\n",
      "    place type_f   type_g or_seq      rel_n       rel_16    rel_ab  log_n  \n",
      "194   ion  water  control    arg   1.000000  2978.516429  0.000336 -3.474  \n",
      "283   ion  water  control    arg   1.503142  2978.516429  0.000505 -3.297  \n",
      "299   ion  water  control    mge   3.491403  2978.516429  0.001172 -2.931  \n",
      "308   ion  water  control    arg  37.068072  2978.516429  0.012445 -1.905  \n",
      "340   ion  water  control    mge   1.148154  2978.516429  0.000385 -3.414  \n"
     ]
    }
   ],
   "source": [
    "big_full_data = pd.DataFrame() # el df final\n",
    "amp_data = full_data.dropna()\n",
    "\n",
    "for sample_num in amp_data[\"sample\"].unique():\n",
    "    # Extraer solo la info de los arg de cada muestra concreta\n",
    "    sample_data = amp_data[(amp_data[\"sample\"] == sample_num)]\n",
    "    \n",
    "    # Juntar ambos. Ojo, las celdas no compartidas no se rellenan\n",
    "    sample_data = pd.merge(arg_names, sample_data, how = \"outer\", on = [\"Assay\"])\n",
    "    \n",
    "    # Sustituir los \"nan\" de las celdas vacías de sample por el sample_num en cuestión, para poder diferenciarlas\n",
    "    sample_data.fillna({\"sample\": sample_num}, inplace = True)\n",
    "    sample_data.fillna({\"Ct\": 0}, inplace = True) #ya que estoy también relleno el Ct\n",
    "    \n",
    "    # Unir el df provisional al denifinitivo y vuelta a empezar\n",
    "    big_full_data = pd.concat([big_full_data, sample_data])\n",
    "\n",
    "print(big_full_data.shape[0])\n",
    "check_df = big_full_data.dropna()\n",
    "print(check_df.tail())\n",
    "del check_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44c4daed",
   "metadata": {},
   "source": [
    "Hay, de nuevo, una columna extra que podemos eliminar para ahorrar memoria, y otra que ha retenido un nombre inadecuado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2ec15ec9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay</th>\n",
       "      <th>antib</th>\n",
       "      <th>sample</th>\n",
       "      <th>Ct</th>\n",
       "      <th>place</th>\n",
       "      <th>type_f</th>\n",
       "      <th>type_g</th>\n",
       "      <th>or_seq</th>\n",
       "      <th>rel_n</th>\n",
       "      <th>rel_16</th>\n",
       "      <th>rel_ab</th>\n",
       "      <th>log_n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aacC2</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>23.78</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>9.246982</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.000146</td>\n",
       "      <td>-3.837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aacA/aphD</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>26.05</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>1.927525</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>-4.518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(6')-II</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>18.80</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>288.403150</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.004539</td>\n",
       "      <td>-2.343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aphA3</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sat4</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>dfrAB4</td>\n",
       "      <td>trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>dfrC</td>\n",
       "      <td>trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>dfrG</td>\n",
       "      <td>trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>dfrK</td>\n",
       "      <td>trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>dfrBmulti</td>\n",
       "      <td>trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9932 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Assay           antib  sample     Ct   place type_f   type_g or_seq  \\\n",
       "0         aacC2  Aminoglycoside    25.0  23.78  ardley    PUR  plastic    arg   \n",
       "1     aacA/aphD  Aminoglycoside    25.0  26.05  ardley    PUR  plastic    arg   \n",
       "2    aac(6')-II  Aminoglycoside    25.0  18.80  ardley    PUR  plastic    arg   \n",
       "3         aphA3  Aminoglycoside    25.0   0.00     NaN    NaN      NaN    NaN   \n",
       "4          sat4  Aminoglycoside    25.0   0.00     NaN    NaN      NaN    NaN   \n",
       "..          ...             ...     ...    ...     ...    ...      ...    ...   \n",
       "377      dfrAB4    trimethoprim    14.0   0.00     NaN    NaN      NaN    NaN   \n",
       "378        dfrC    trimethoprim    14.0   0.00     NaN    NaN      NaN    NaN   \n",
       "379        dfrG    trimethoprim    14.0   0.00     NaN    NaN      NaN    NaN   \n",
       "380        dfrK    trimethoprim    14.0   0.00     NaN    NaN      NaN    NaN   \n",
       "381   dfrBmulti    trimethoprim    14.0   0.00     NaN    NaN      NaN    NaN   \n",
       "\n",
       "          rel_n        rel_16    rel_ab  log_n  \n",
       "0      9.246982  63533.093185  0.000146 -3.837  \n",
       "1      1.927525  63533.093185  0.000030 -4.518  \n",
       "2    288.403150  63533.093185  0.004539 -2.343  \n",
       "3           NaN           NaN       NaN    NaN  \n",
       "4           NaN           NaN       NaN    NaN  \n",
       "..          ...           ...       ...    ...  \n",
       "377         NaN           NaN       NaN    NaN  \n",
       "378         NaN           NaN       NaN    NaN  \n",
       "379         NaN           NaN       NaN    NaN  \n",
       "380         NaN           NaN       NaN    NaN  \n",
       "381         NaN           NaN       NaN    NaN  \n",
       "\n",
       "[9932 rows x 12 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "big_full_data.drop(columns = [\"antib_y\"], axis = 1, inplace = True)\n",
    "big_full_data.rename(columns = {\"antib_x\": \"antib\"}, inplace = True)\n",
    "\n",
    "big_full_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85948719",
   "metadata": {},
   "source": [
    "Y por cómo está programado, se hace necesario volver a asignar los valores de clasificación antes puestos (pues aunque no haya valores de Ct, sí que interesa saber qué tipo de muestras son)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ebaf525f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay</th>\n",
       "      <th>antib</th>\n",
       "      <th>sample</th>\n",
       "      <th>Ct</th>\n",
       "      <th>place</th>\n",
       "      <th>type_f</th>\n",
       "      <th>type_g</th>\n",
       "      <th>or_seq</th>\n",
       "      <th>rel_n</th>\n",
       "      <th>rel_16</th>\n",
       "      <th>rel_ab</th>\n",
       "      <th>log_n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aacC2</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>23.78</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>9.246982</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.000146</td>\n",
       "      <td>-3.837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aacA/aphD</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>26.05</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>1.927525</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>-4.518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(6')-II</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>18.80</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>288.403150</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.004539</td>\n",
       "      <td>-2.343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aphA3</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sat4</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>dfrAB4</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>dfrC</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>dfrG</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>dfrK</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>dfrBmulti</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9932 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Assay           antib  sample     Ct   place type_f   type_g or_seq  \\\n",
       "0         aacC2  Aminoglycoside    25.0  23.78  ardley    PUR  plastic    arg   \n",
       "1     aacA/aphD  Aminoglycoside    25.0  26.05  ardley    PUR  plastic    arg   \n",
       "2    aac(6')-II  Aminoglycoside    25.0  18.80  ardley    PUR  plastic    arg   \n",
       "3         aphA3  Aminoglycoside    25.0   0.00  ardley    PUR  plastic    arg   \n",
       "4          sat4  Aminoglycoside    25.0   0.00  ardley    PUR  plastic    arg   \n",
       "..          ...             ...     ...    ...     ...    ...      ...    ...   \n",
       "377      dfrAB4    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "378        dfrC    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "379        dfrG    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "380        dfrK    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "381   dfrBmulti    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "\n",
       "          rel_n        rel_16    rel_ab  log_n  \n",
       "0      9.246982  63533.093185  0.000146 -3.837  \n",
       "1      1.927525  63533.093185  0.000030 -4.518  \n",
       "2    288.403150  63533.093185  0.004539 -2.343  \n",
       "3           NaN           NaN       NaN    NaN  \n",
       "4           NaN           NaN       NaN    NaN  \n",
       "..          ...           ...       ...    ...  \n",
       "377         NaN           NaN       NaN    NaN  \n",
       "378         NaN           NaN       NaN    NaN  \n",
       "379         NaN           NaN       NaN    NaN  \n",
       "380         NaN           NaN       NaN    NaN  \n",
       "381         NaN           NaN       NaN    NaN  \n",
       "\n",
       "[9932 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#primero, definir puntos de muestreo\n",
    "big_full_data.loc[(big_full_data[\"sample\"] < 13), \"place\"] = \"uru\"\n",
    "big_full_data.loc[(big_full_data[\"sample\"] > 12) & (big_full_data[\"sample\"] < 19), \"place\"] = \"ion\"\n",
    "big_full_data.loc[(big_full_data[\"sample\"] > 18), \"place\"] = \"ardley\"\n",
    "\n",
    "#segundo, definir tipo fino\n",
    "big_full_data.loc[(big_full_data[\"sample\"].isin([1, 2, 3, 13, 19, 20, 21])), \"type_f\"] = \"soil\"\n",
    "big_full_data.loc[(big_full_data[\"sample\"].isin([4, 5, 6, 14])), \"type_f\"] = \"water\"\n",
    "big_full_data.loc[(big_full_data[\"sample\"].isin([7, 8, 9, 15, 16, 22, 23, 24])), \"type_f\"] = \"EPS\"\n",
    "big_full_data.loc[(big_full_data[\"sample\"].isin([10, 11, 12, 17, 18, 25, 26, 27])), \"type_f\"] = \"PUR\"\n",
    "\n",
    "#tercero, tipo general\n",
    "big_full_data.loc[(big_full_data[\"type_f\"] == \"EPS\") | (big_full_data[\"type_f\"] == \"PUR\"), \"type_g\"] = \"plastic\"\n",
    "big_full_data.loc[(big_full_data[\"type_f\"] == \"water\") | (big_full_data[\"type_f\"] == \"soil\"), \"type_g\"] = \"control\"\n",
    "\n",
    "#cuarto, corregir OTRA VEZ los nombres \"defectuosos\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"MDR-chromo\") | (big_full_data[\"antib\"] == \"MDR-mobile\"), \"antib\"] = \"MDR\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"Beta lactamase\") | (big_full_data[\"antib\"] == \"beta Lactam\"), \"antib\"] = \"Beta Lactam\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"trimethoprim\"), \"antib\"] = \"Trimethoprim\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"Tetracycline \"), \"antib\"] = \"Tetracycline\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"fluoroquinolone\"), \"antib\"] = \"Fluoroquinolone\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"other\"), \"antib\"] = \"Other\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"Insertional \"), \"antib\"] = \"Insertional\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"Plasmid-inc \"), \"antib\"] = \"Plasmid-inc\"\n",
    "big_full_data.loc[(big_full_data[\"antib\"] == \"plasmid\"), \"antib\"] = \"Plasmid\"\n",
    "\n",
    "#y ya el tipo de secuencia\n",
    "big_full_data.loc[(big_full_data[\"antib\"].isin(arg_lst)), \"or_seq\"] = \"arg\"\n",
    "big_full_data.loc[(~big_full_data[\"antib\"].isin(arg_lst)), \"or_seq\"] = \"mge\"\n",
    "\n",
    "big_full_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "882faaa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ya se puede pasar a csv\n",
    "big_full_data.to_csv(\"../data/resistome_data/clean_data/ab_data_all.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1ff32e2-3c65-43ad-9f5d-15ecb7122b61",
   "metadata": {},
   "source": [
    "---\n",
    "Necesito salir de un apuro mientras esperamos a que desde Francia nos manden unas cosas. Asi que voy a crearme un dataset falseado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bc35fc57-a8e8-423c-b082-d1fd385f08dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "arnr_vals[21] = (arnr_vals[19] + arnr_vals[20])/2\n",
    "amp_data = full_data[(~full_data[\"antib\"].isna())]\n",
    "amp_data = amp_data[(~amp_data[\"sample\"].isna())]\n",
    "\n",
    "for sample_num,rel_16 in arnr_vals.items():\n",
    "    amp_data.loc[(amp_data[\"sample\"] == sample_num), \"rel_16\"] = rel_16 \n",
    "amp_data[\"rel_ab\"] = amp_data[\"rel_n\"]/amp_data[\"rel_16\"] \n",
    "\n",
    "\n",
    "amp_data[\"log_n\"] = np.log10(amp_data[\"rel_ab\"])\n",
    "amp_data.loc[(amp_data[\"rel_ab\"] == 0.0), \"log_n\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b32c4a96-0a54-4626-a838-adbd0c64e5af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay</th>\n",
       "      <th>antib</th>\n",
       "      <th>sample</th>\n",
       "      <th>Ct</th>\n",
       "      <th>place</th>\n",
       "      <th>type_f</th>\n",
       "      <th>type_g</th>\n",
       "      <th>or_seq</th>\n",
       "      <th>rel_n</th>\n",
       "      <th>rel_16</th>\n",
       "      <th>rel_ab</th>\n",
       "      <th>log_n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aacC2</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>23.78</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>9.246982</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.000146</td>\n",
       "      <td>-3.837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aacA/aphD</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>26.05</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>1.927525</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>-4.518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(6')-II</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>18.80</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>288.403150</td>\n",
       "      <td>63533.093185</td>\n",
       "      <td>0.004539</td>\n",
       "      <td>-2.343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aphA3</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sat4</td>\n",
       "      <td>Aminoglycoside</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ardley</td>\n",
       "      <td>PUR</td>\n",
       "      <td>plastic</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>dfrAB4</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>dfrC</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>dfrG</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>dfrK</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>dfrBmulti</td>\n",
       "      <td>Trimethoprim</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ion</td>\n",
       "      <td>water</td>\n",
       "      <td>control</td>\n",
       "      <td>arg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10314 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Assay           antib  sample     Ct   place type_f   type_g or_seq  \\\n",
       "0         aacC2  Aminoglycoside    25.0  23.78  ardley    PUR  plastic    arg   \n",
       "1     aacA/aphD  Aminoglycoside    25.0  26.05  ardley    PUR  plastic    arg   \n",
       "2    aac(6')-II  Aminoglycoside    25.0  18.80  ardley    PUR  plastic    arg   \n",
       "3         aphA3  Aminoglycoside    25.0   0.00  ardley    PUR  plastic    arg   \n",
       "4          sat4  Aminoglycoside    25.0   0.00  ardley    PUR  plastic    arg   \n",
       "..          ...             ...     ...    ...     ...    ...      ...    ...   \n",
       "377      dfrAB4    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "378        dfrC    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "379        dfrG    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "380        dfrK    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "381   dfrBmulti    Trimethoprim    14.0   0.00     ion  water  control    arg   \n",
       "\n",
       "          rel_n        rel_16    rel_ab  log_n  \n",
       "0      9.246982  63533.093185  0.000146 -3.837  \n",
       "1      1.927525  63533.093185  0.000030 -4.518  \n",
       "2    288.403150  63533.093185  0.004539 -2.343  \n",
       "3           NaN           NaN       NaN    NaN  \n",
       "4           NaN           NaN       NaN    NaN  \n",
       "..          ...           ...       ...    ...  \n",
       "377         NaN           NaN       NaN    NaN  \n",
       "378         NaN           NaN       NaN    NaN  \n",
       "379         NaN           NaN       NaN    NaN  \n",
       "380         NaN           NaN       NaN    NaN  \n",
       "381         NaN           NaN       NaN    NaN  \n",
       "\n",
       "[10314 rows x 12 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "big_false_data = pd.DataFrame() # el df final\n",
    "\n",
    "for sample_num in amp_data[\"sample\"].unique():\n",
    "    sample_data = amp_data[(amp_data[\"sample\"] == sample_num)]\n",
    "    sample_data = pd.merge(arg_names, sample_data, how = \"outer\", on = [\"Assay\"])\n",
    "    sample_data.fillna({\"sample\": sample_num}, inplace = True)\n",
    "    sample_data.fillna({\"Ct\": 0}, inplace = True)\n",
    "    big_false_data = pd.concat([big_false_data, sample_data])\n",
    "\n",
    "big_false_data.drop(columns = [\"antib_y\"], axis = 1, inplace = True)\n",
    "big_false_data.rename(columns = {\"antib_x\": \"antib\"}, inplace = True)\n",
    "\n",
    "#primero, definir puntos de muestreo\n",
    "big_false_data.loc[(big_false_data[\"sample\"] < 13), \"place\"] = \"uru\"\n",
    "big_false_data.loc[(big_false_data[\"sample\"] > 12) & (big_false_data[\"sample\"] < 19), \"place\"] = \"ion\"\n",
    "big_false_data.loc[(big_false_data[\"sample\"] > 18), \"place\"] = \"ardley\"\n",
    "\n",
    "#segundo, definir tipo fino\n",
    "big_false_data.loc[(big_false_data[\"sample\"].isin([1, 2, 3, 13, 19, 20, 21])), \"type_f\"] = \"soil\"\n",
    "big_false_data.loc[(big_false_data[\"sample\"].isin([4, 5, 6, 14])), \"type_f\"] = \"water\"\n",
    "big_false_data.loc[(big_false_data[\"sample\"].isin([7, 8, 9, 15, 16, 22, 23, 24])), \"type_f\"] = \"EPS\"\n",
    "big_false_data.loc[(big_false_data[\"sample\"].isin([10, 11, 12, 17, 18, 25, 26, 27])), \"type_f\"] = \"PUR\"\n",
    "\n",
    "#tercero, tipo general\n",
    "big_false_data.loc[(big_false_data[\"type_f\"] == \"EPS\") | (big_false_data[\"type_f\"] == \"PUR\"), \"type_g\"] = \"plastic\"\n",
    "big_false_data.loc[(big_false_data[\"type_f\"] == \"water\") | (big_false_data[\"type_f\"] == \"soil\"), \"type_g\"] = \"control\"\n",
    "\n",
    "#cuarto, corregir OTRA VEZ los nombres \"defectuosos\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"MDR-chromo\") | (big_false_data[\"antib\"] == \"MDR-mobile\"), \"antib\"] = \"MDR\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"Beta lactamase\") | (big_false_data[\"antib\"] == \"beta Lactam\"), \"antib\"] = \"Beta Lactam\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"trimethoprim\"), \"antib\"] = \"Trimethoprim\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"Tetracycline \"), \"antib\"] = \"Tetracycline\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"fluoroquinolone\"), \"antib\"] = \"Fluoroquinolone\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"other\"), \"antib\"] = \"Other\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"Insertional \"), \"antib\"] = \"Insertional\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"Plasmid-inc \"), \"antib\"] = \"Plasmid-inc\"\n",
    "big_false_data.loc[(big_false_data[\"antib\"] == \"plasmid\"), \"antib\"] = \"Plasmid\"\n",
    "\n",
    "#y ya el tipo de secuencia\n",
    "big_false_data.loc[(big_false_data[\"antib\"].isin(arg_lst)), \"or_seq\"] = \"arg\"\n",
    "big_false_data.loc[(~big_false_data[\"antib\"].isin(arg_lst)), \"or_seq\"] = \"mge\"\n",
    "\n",
    "big_false_data[\"sample\"].unique()\n",
    "big_false_data.to_csv(\"../data/resistome_data/clean_data/ab_data_false.csv\")\n",
    "big_false_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
